lesson1/contracts/Token.sol

```diff
@@ -0,0 +1,102 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+//import "hardhat/console.sol";
+
+contract Token {
+    event Transfer(address indexed _from, address indexed _to, uint256 _value);
+    event Approval(
+        address indexed _owner,
+        address indexed _spender,
+        uint256 _value
+    );
+
+    struct ApprovalStruct {
+        address owner;
+        address approved_address;
+        uint256 approved_amount;
+    }
+
+    mapping(address => uint256) balances;
+    mapping(address => mapping(address => ApprovalStruct)) approvals;
+
+    constructor() {
+        balances[msg.sender] = totalSupply();
+        emit Transfer(address(0), msg.sender, totalSupply());
+    }
+
+    function name() public view returns (string memory) {
+        return "Leocode Token";
+    }
+
+    function symbol() public view returns (string memory) {
+        return "LEO";
+    }
+
+    function decimals() public view returns (uint8) {
+        return 18;
+    }
+
+    function totalSupply() public view returns (uint256) {
+        uint256 base_decimals_multiplier = 10**decimals();
+        return 100_000 * base_decimals_multiplier;
+    }
+
+    function balanceOf(address _owner) public view returns (uint256 balance) {
+        return balances[_owner];
+    }
+
+    function transfer(address _to, uint256 _value)
+        public
+        returns (bool success)
+    {
+        require(_to != address(0), "Can not transfer token to address 0");
+        require(balances[msg.sender] >= _value, "Not enough tokens");
```

Comment:
Wg tego require jeśli ktoś ma dokładnie tyle tokenów, ile trzeba, to przelew i tak się nie uda ;)